### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by prakashsakari**
This PR has the following
1. User Authentication with persistent login
2. Private Routes
3. API integration for different features

Preview Link: https://deploy-preview-18--videoporium.netlify.app/

### CODE REVIEW

**Review by Rohitprasad83 — APPROVED** [review]
Clean code, looks good to merge

File: src/App.js [hunk 1]
```diff
@@ -1,15 +1,18 @@
 import "./App.css";
-import { Home, SingleVideo, Liked, WatchLater, History } from "./pages";
+import {RequiresAuth} from "./components";
+import { Home, SingleVideo, Liked, WatchLater, History, Login, SignUp } from "./pages";
 import { Routes, Route } from "react-router-dom";
 
 export default function App() {
```
Comment (ravikumar0403) [inline src/App.js:6]: You can change this to use es6 syntax of arrow functions

File: src/components/Auth/Login.jsx [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                    <button className="button btn-link-primary cursor mg-left">
+                        <Link className="link-primary" to="/login">
+                            Forgot your Password?
+                        </Link>
+                    </button>
+                </div>
+                <div className="cta">
+                    <button
+                        className="login-btn button btn-primary cursor btn-margin sign-up-btn"
+                        onClick={() => {
```
Comment (ravikumar0403) [inline src/components/Auth/Login.jsx]: You can move to logic to a function in order to keep jsx and functional logic separate

File: src/context/auth-context.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
+import { createContext, useContext, useState, useReducer } from "react";
+import axios from "axios";
+import { useEffect} from "react";
```
Comment (ravikumar0403) [inline src/context/auth-context.js]: This can also be imported in the first line itself

File: src/context/auth-context.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    const logOutHandler = () => {
+        if (eToken){
+            setEToken("");
+            setEUser("");
+            localStorage.removeItem("token");
+            localStorage.removeItem("likes");
+            localStorage.removeItem("watchlater");
+            localStorage.removeItem("user");
+            localStorage.removeItem("history");
```
Comment (ravikumar0403) [inline src/context/auth-context.js]: Use `localStorage.clear()` to clear all local storage data

File: src/context/playlist-context.js [hunk 5]
```diff
@@ -1,10 +1,57 @@
...
+      const history = JSON.parse(localStorage.getItem("history" ?? ""));
+      history && history.length > 0 && history.map(video => playlistDispatch({
+        type: "HISTORY",
+        payload: video
+      }));
+    }
+  }, [])
+
+  useEffect(() => {
+    if (localStorage.getItem("option" ?? "")){
```
Comment (ravikumar0403) [inline src/context/playlist-context.js]: `??` is not required here as null or empty string both result to falsy value. I dont understand the purpose of using nullish coalescing operator here

**Review by ravikumar0403 — APPROVED** [review]

File: src/components/Auth/Login.jsx (hunk 2 above)
Comment (prakashsakari) [reply src/components/Auth/Login.jsx]: thank you ravi, updated the changes

File: src/context/auth-context.js (hunk 3 above)
Comment (prakashsakari) [reply src/context/auth-context.js]: thank you for pointing this out ravi, updated changes

File: src/context/auth-context.js (hunk 4 above)
Comment (prakashsakari) [reply src/context/auth-context.js]: thank you for this ravi, updated changes

File: src/context/playlist-context.js (hunk 5 above)
Comment (prakashsakari) [reply src/context/playlist-context.js]: I was trying to handle an edge case, but now I have made the changes and I think with nullish coalescing empty string is not a falsy value

File: src/App.js (hunk 1 above)
Comment (prakashsakari) [reply src/App.js:6]: will look into this, thanks ravi

**Pull request merged**